### PR TITLE
Apply design

### LINKED ISSUE #156: Apply
As discussed on lambdaLille, `Apply` (Applicative without `pure`) can be useful. 
(The full API should exposes every combinators that can be written without `pure`).

### PR DESCRIPTION
**Pull request by d-plaindoux**
- [X] Naming convention adding pure (implicit in the previous version)
- [X] Apply specification
- [x] Apply implementation
- [x] Express Applicative thanks to Apply
- [x] Apply QuickCheck
- [x] Applicative From Apply
- [x] Documentation

Fix #156

### CODE REVIEW

**Comment by gr-im** [comment]
If it may help. Function that be provided by Apply shoud be: 
- `apply and <*>`
- `*> and <*` and it could be an opportunity for having not-infix version (`apply_fst` and `apply_snd`)
- every `lift` function

File: lib/preface_make/applicative.mli [hunk 1]
```diff
@@ -115,14 +115,16 @@ module Via
...
+module Core_via_pure_and_apply
+    (Req : Preface_specs.Applicative.WITH_PURE_AND_APPLY) :
   Preface_specs.Applicative.CORE with type 'a t = 'a Req.t
 
-module Core_via_lift2 (Req : Preface_specs.Applicative.WITH_LIFT2) :
+module Core_via_pure_and_lift2
+    (Req : Preface_specs.Applicative.WITH_PURE_AND_LIFT2) :
   Preface_specs.Applicative.CORE with type 'a t = 'a Req.t
 
 (** {2 Deriving Operation} *)
```
Comment (gr-im) [inline lib/preface_make/applicative.mli:130]: I think a module should be added: 
```ocaml
module Over_apply (A: Preface.Specs.APPLY) (REQ: Preface.Specs.Applicative.WITH_PURE) : ...
```

File: lib/preface_make/apply.ml [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+  include Via_map_and_product (struct
+    type 'a t = 'a Applicative.t
+
+    let map = Applicative.map
+
+    let product a b =
+      Applicative.apply (Applicative.map (fun a b -> (a, b)) a) b
+    ;;
+  end)
+end
```
Comment (gr-im) [inline lib/preface_make/apply.ml:123]: I think that, if all `Applicative` is an `Apply` (which it should be), this module should be rewritten like `From_alternative`:
```ocaml
module From_applicative (Applicative : Preface_specs.APPLICATIVE) = Applicative
```

File: lib/preface_make/apply.ml [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+struct
+  module Core = Core_over_functor_via_lift2 (Req) (Functor)
+  module Operation = Operation (Core)
+  module Syntax = Syntax (Core)
+  module Infix = Infix (Core) (Operation)
+  include Core
+  include Operation
+  include Syntax
+  include Infix
+end
```
Comment (gr-im) [inline lib/preface_make/apply.ml:111]: Um, the API is not consistent with `Via_map_and_product`. I suggest to have several modules: 

- `Over_functor_via_product`
- `Via_map_and_product`
- `Over_functor_via_apply`
- `Via_map_and_apply`
- `Over_functor_via_lift2`
- `Via_map_and_lift2`

(And the same for the versions of `Core`).

Note that I wonder if it wouldn't be better to do something like this: 

```ocaml
module Over_functor (F: Preface_specs.Functor.CORE) : sig 
   module Via_apply (...)
   (* etc *)
end
```
But that would mean a "big change in the whole Preface library...

File: lib/preface_make/apply.ml [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+module Core_over_functor_via_lift2
+    (Req : Preface_specs.Apply.WITH_LIFT2)
+    (Functor : Preface_specs.Functor.WITH_MAP with type 'a t = 'a Req.t) =
+struct
+  include Req
+  include Functor
+
+  let apply f a = lift2 (fun x -> x) f a
+  let product a b = apply (map (fun a b -> (a, b)) a) b
+end
```
Comment (gr-im) [inline lib/preface_make/apply.ml:31]: The API is not consistent, CF my comment below.

**Review by gr-im — CHANGES_REQUESTED** [review]
Thank you very much for the work. 
I think that, as for Semigroupoidal, we should add the tests of laws (which would be like taking part of the tests of Applicative).

File: lib/preface_make/applicative.mli (hunk 1 above)
Comment (d-plaindoux) [reply lib/preface_make/applicative.mli:143]: Done

File: lib/preface_make/apply.ml (hunk 2 above)
Comment (d-plaindoux) [reply lib/preface_make/apply.ml:138]: Done

File: lib/preface_make/apply.ml (hunk 3 above)
Comment (d-plaindoux) [reply lib/preface_make/apply.ml]: Done

File: lib/preface_make/apply.ml (hunk 4 above)
Comment (d-plaindoux) [reply lib/preface_make/apply.ml]: Done

File: test/preface_laws/applicative.ml [hunk 5]
```diff
@@ -1,6 +1,6 @@
 open Preface_qcheck
 
-module Perserve_identity
```
Comment (gr-im) [inline test/preface_laws/applicative.ml:3]: Well spotted ;)

**Review by gr-im — APPROVED** [review]
Thanks!

**Pull request merged**